### DISCUSSION TITLE
Submacro in Tiddlywiki

### DISCUSSION THREAD

**Discussion by kookma** [TiddlyWiki/TiddlyWiki5#6287] [Q&A]
A nested macro pragma is allowed in Tiddlywiki! See below

```
\define main(p1,p2)
\define subm(n) Hello $n$, my dear friend!
<$list filter="[[$p1$]] [[$p2$]]">
<$macrocall $name=subm n=<<currentTiddler>> />
</$list>
\end
```

But I cannot use variable in main macro from sub macro: like `$(name)$`  in the below example

```
\define main(p1,p2)
\define subm() Hello $(name)$, my dear friend!
<$list filter="[[$p1$]] [[$p2$]]" variable=name>
<<subm>>
</$list>
\end
```

**Questions**:

1. do you recommend such sub-macro?
2. what is the scope of variable in main and sub macro?

**Comment by pmario** [comment]
Hi, 
I know that the code doesn't work, but the `<<main x y>>` call is missing. ... Devs also like copy / pasting, because it saves time ;)

The outer macro starts a new parser instance which saves the inner macro as a text string. So whenever the inner macro is called it needs to start the parser again. ... 

The parse-tree of the outer macro can be cached, if we use `[<__p1__> <__p2__>]` instead of `$p1$` ... I think the inner macro will never be cached. But I'm not 100% sure here. Would need to have a look at the code. 

The following code works, but I personally wouldn't use macros that way. IMO it makes the code even more complex for others to read and may be debug. 

```
\define main(p1,p2)
\define subm() Hello <<name>>, my dear friend!
<$list filter="[<__p1__>] [<__p2__>]" variable=name>
<<subm>>
</$list>
\end

<<main a b>>
```

**Comment by kookma** [comment]
Thank you @pmario 
So, the submacro in this way are not efficient and your recommendation is to have separate macros!

**Reply by pmario** [reply]
They are separate macros. .. 

IMO it's about visibility and complexity

 - If you tag the tiddler $:/tags/Macro .. "main" will be global but "subm" will not be seen globally. Only "main" sees it. ... 
 - I think some users may be confused what's going on here. .. 

On the other hand the "encapsulation" may be an advantage ... 

As I wrote. I don't know if the inner macro is cached. So it depends on what you do, how much code it contains _and_ how often it is executed. 

So if it will be used for long list-widget outputs there may be a problem. ... If it is for a form-tiddler everything should be just fine. 

If you use it in a tiddler that is not tagged $:/tags/Macro the overhead should be almost similar to a standard macro call. 

We would need to test it with 10k to 30k tiddlers and long lists. ... IMO mobile phones are the main concern atm, most PCs are fast enough already

**Reply by pmario** [reply]
With long lists, the most time consuming code is creating and updating the DOM and not the wikitext code interpretation
